src/org/pentaho/reporting/platform/plugin/async/IAsyncReportState.java

```diff
@@ -29,6 +29,24 @@
   int getProgress();
 
   /**
+   *
+   * @return Page is currently being processed
+   */
+  int getPage();
+
+  /**
+   *
+   * @return Row is currently being processed
+   */
+  int getRow();
+
```

Comment:
Instead of the row, which is useless for showing a progress bar, compute a percentage value that you can give to the GUI. This way the progress bar can go from empty to full as the report processes.

This means you need to take both row and max-row into account. Also remember: We have multiple passes, so a percentage based purely on result = row / maxRow will go from 0 to 100 and back to zero on each pass.